cpp/src/gandiva/interval_holder.cc

```diff
@@ -0,0 +1,277 @@
+    return GetIntervalDayFromWeeks(period_in_weeks, out_valid);
+  }
+
+  std::string days_in_period;
+  std::string hours_in_period;
+  std::string minutes_in_period;
+  std::string seconds_in_period;
+  std::string ignored_string;  // string to store unnecessary captured groups
+  if (std::regex_match(data, period_not_contains_time)) {
+    if (RE2::FullMatch(data, iso8601_period_without_time, &ignored_string,
+                       &ignored_string, &days_in_period)) {
+      return GetIntervalDayFromCompletePeriod(days_in_period, hours_in_period,
+                                              minutes_in_period, seconds_in_period,
+                                              out_valid);
+    }
+
+    return_error(ctx, data);
+    return 0;
+  }
+
+  if (RE2::FullMatch(data, iso8601_complete_period, &ignored_string, &ignored_string,
+                     &days_in_period, &hours_in_period, &minutes_in_period,
+                     &seconds_in_period)) {
+    return GetIntervalDayFromCompletePeriod(
+        days_in_period, hours_in_period, minutes_in_period, seconds_in_period, out_valid);
+  }
+
+  return_error(ctx, data);
+  return 0;
+}
+
+Status IntervalDaysHolder::Make(const FunctionNode& node,
+                                std::shared_ptr<IntervalDaysHolder>* holder) {
+  const std::string function_name("castINTERVALDAY");
+  return IntervalHolder<IntervalDaysHolder>::Make(node, holder, function_name);
+}
+
+Status IntervalDaysHolder::Make(int32_t suppress_errors,
+                                std::shared_ptr<IntervalDaysHolder>* holder) {
+  return IntervalHolder<IntervalDaysHolder>::Make(suppress_errors, holder);
+}
+
+Status IntervalYearsHolder::Make(const FunctionNode& node,
+                                 std::shared_ptr<IntervalYearsHolder>* holder) {
+  const std::string function_name("castINTERVALYEAR");
+  return IntervalHolder<IntervalYearsHolder>::Make(node, holder, function_name);
+}
+
+Status IntervalYearsHolder::Make(int32_t suppress_errors,
+                                 std::shared_ptr<IntervalYearsHolder>* holder) {
+  return IntervalHolder<IntervalYearsHolder>::Make(suppress_errors, holder);
+}
+
+// The operator will cast a generic string defined by the user into an interval of months.
+// There are two formats of strings that are acceptable:
+//   - The period in millis: '238398430'
+//   - The period using a ISO8601 compatible format: 'P[n]Y[n]M[n]DT[n]H[n]M[n]S or P[n]W'
+int32_t IntervalYearsHolder::operator()(ExecutionContext* ctx, const std::string& data,
+                                        bool in_valid, bool* out_valid) {
+  *out_valid = false;
+
+  if (!in_valid) {
+    return 0;
+  }
+
+  if (RE2::FullMatch(data, period_only_contains_numbers)) {
+    return GetIntervalYearFromNumber(data, out_valid);
+  }
+
+  std::string period_in_weeks;
+  if (RE2::FullMatch(data, iso8601_period_with_weeks, &period_in_weeks)) {
+    *out_valid = true;
```

Comment:
The reason for the existence of that branch is because [in ISO](https://en.wikipedia.org/wiki/ISO_8601#Durations) the user must choose to represent a period using one of these two ways:

- `PnYnMTnHnM`
- `PnW`

So if the user represents using the second format, we will process the value in a different way.

Reply:
and are you saying that if we specify period in weeks, the number of months should be always 0, as per rfc. In that case I understand this branch now.